File: audio_loop.go
package main

import (
	"encoding/binary"
	"io"
	"log"
	"os/exec"
	"strconv"

	"ai_box/aec"

	vado "github.com/maxhawkins/go-webrtc-vad"
)

func audioLoop(aecProc *aec.Processor, vadEng *vado.VAD) {
	cmd := exec.Command("arecord",
		"-D", arecordDevice,
		"-c", strconv.Itoa(arecordChannels),
		"-r", strconv.Itoa(arecordRate),
		"-f", "S16_LE",
		"-t", "raw",
		"--period-size="+strconv.Itoa(arecordPeriodSize),
		"--buffer-size="+strconv.Itoa(arecordBufferSize),
	)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		log.Fatal(err)
	}
	if err := cmd.Start(); err != nil {
		log.Fatal(err)
	}
	log.Println("🎤 麦克风已开启...")

	readBuf := make([]byte, 256*10*2)
	vadAccumulator := make([]int16, 0, 1024)
	var asrBuffer []int16
	silenceCount, speechCount := 0, 0
	triggered := false
	ducked := false
	fallbackMono := make([]int16, 256)

	for {
		if _, err := io.ReadFull(stdout, readBuf); err != nil {
			break
		}
		rawInt16 := make([]int16, 256*10)
		for i := 0; i < len(rawInt16); i++ {
			rawInt16[i] = int16(binary.LittleEndian.Uint16(readBuf[i*2 : i*2+2]))
		}
		clean, _ := aecProc.Process(rawInt16)
		if clean == nil {
			// AEC 异常回退：取第 0 通道直通，避免整段音频被丢弃导致“说了却识别不到”
			for i := 0; i < 256; i++ {
				fallbackMono[i] = rawInt16[i*10+0]
			}
			clean = fallbackMono
		}
		vadAccumulator = append(vadAccumulator, clean...)

		for len(vadAccumulator) >= 320 {
			frame := vadAccumulator[:320]
			vadAccumulator = vadAccumulator[320:]
			vadBuf := make([]byte, 640)
			for i, v := range frame {
				binary.LittleEndian.PutUint16(vadBuf[i*2:], uint16(v))
			}
			active, _ := vadEng.Process(16000, vadBuf)
			if active {
				speechCount++
				silenceCount = 0
			} else {
				silenceCount++
				speechCount = 0
			}

			// 先快速 Duck（听感上立刻压低背景音），再决定是否进入 ASR 录音段
			if speechCount > 2 && !ducked {
				ducked = true
				musicMgr.Duck()
			}

			if speechCount > 10 && !triggered {
				triggered = true
			}
			if triggered {
				asrBuffer = append(asrBuffer, frame...)
				if silenceCount > 10 || len(asrBuffer) > 16000*8 {
					if len(asrBuffer) > 4800 {
						finalData := make([]int16, len(asrBuffer))
						copy(finalData, asrBuffer)
						go processASR(finalData)
					} else {
						musicMgr.Unduck()
					}
					asrBuffer = []int16{}
					triggered = false
					ducked = false
					silenceCount = 0
				}
			} else {
				if len(asrBuffer) > 8000 {
					asrBuffer = asrBuffer[320:]
				}
				asrBuffer = append(asrBuffer, frame...)
			}
		}
	}
}

File: py_ai_box/audio_loop.py
import subprocess
import threading
from array import array

from . import state
from .aec import AECProcessor, FRAME_SIZE, INPUT_TOTAL_CH
from . import config_runtime
from .asr import process_asr


def audio_loop(aec_proc: AECProcessor, vad_engine) -> None:
    cmd = [
        "arecord",
        "-D", config_runtime.arecord_device,
        "-c", str(config_runtime.arecord_channels),
        "-r", str(config_runtime.arecord_rate),
        "-f", "S16_LE",
        "-t", "raw",
        f"--period-size={config_runtime.arecord_period_size}",
        f"--buffer-size={config_runtime.arecord_buffer_size}",
    ]
    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
    if proc.stdout is None:
        raise RuntimeError("arecord 启动失败")
    print("🎤 麦克风已开启...")

    read_bytes = FRAME_SIZE * INPUT_TOTAL_CH * 2
    vad_acc = array("h")
    asr_buffer = array("h")
    silence_count = 0
    speech_count = 0
    triggered = False
    ducked = False
    fallback_mono = array("h", [0] * FRAME_SIZE)

    while True:
        if state.shutdown_event.is_set():
            break
        data = proc.stdout.read(read_bytes)
        if not data or len(data) < read_bytes:
            break
        raw = array("h")
        raw.frombytes(data)

        clean, _ = aec_proc.process(raw)
        if clean is None:
            for i in range(FRAME_SIZE):
                fallback_mono[i] = raw[i * INPUT_TOTAL_CH + 0]
            clean = fallback_mono

        vad_acc.extend(clean)

        while len(vad_acc) >= 320:
            frame = vad_acc[:320]
            del vad_acc[:320]

            active = vad_engine.is_speech(frame)
            if active:
                speech_count += 1
                silence_count = 0
            else:
                silence_count += 1
                speech_count = 0

            if speech_count > 2 and not ducked and state.music_mgr is not None:
                ducked = True
                state.music_mgr.duck()

            if speech_count > 10 and not triggered:
                triggered = True

            if triggered:
                asr_buffer.extend(frame)
                if silence_count > 10 or len(asr_buffer) > 16000 * 8:
                    if len(asr_buffer) > 4800:
                        final_data = array("h", asr_buffer)
                        threading.Thread(target=process_asr, args=(final_data,), daemon=True).start()
                    else:
                        if state.music_mgr is not None:
                            state.music_mgr.unduck()
                    asr_buffer = array("h")
                    triggered = False
                    ducked = False
                    silence_count = 0
            else:
                if len(asr_buffer) > 8000:
                    del asr_buffer[:320]
                asr_buffer.extend(frame)
